### ISSUE TITLE
metricsql: support join of two or more expressions in one result

### ISSUE THREAD

**Issue by hagen1778** [VictoriaMetrics/VictoriaMetrics#4372]
### Is your feature request related to a problem? Please describe

MetricsQL, as PromQL, already does intersection of vectors.
For example, `q1 and q2` will return all time series from q1 which intersect with q2.

But for some cases, it is needed to get a union of q1 and q2 in one go.
For example, to get all instances with CPU and mem bigger than 10 cores and 100Gi respectively.
The naive `cpu_count > 10 and mem_bytes > 100Gi` will return `cpu_count` only. If we do union `cpu_count > 10 union mem_bytes > 100Gi` - we'll get those instances which have cpu_count > 10 but mem_bytes lower than 100Gi, and vice versa.

### Describe the solution you'd like

Would be nice to have a new operator to perform intersection of q1 and q2 - `union_distinct`:
```
union_distinct(cpu_count > 10, mem_bytes > 100G)
```
The query above should run in parallel q1 and q2, and return result consisting of intersected results for both of queries (without __name__).

**Comment by valyala** [comment]
The [`union`](https://docs.victoriametrics.com/metricsql/#union) function works in the way you described. For example, the following [MetricsQL]() query returns `cpu_count` metrics with values bigger than 10 plus `mem_bytes` metrics with values bigger than 100G:

```metricsql
union(cpu_count > 10, mem_bytes > 100G)
```

Try [this example at VictoriaMetrics playground](https://play.victoriametrics.com/select/0/vmui/?#/?g0.expr=union%28process_resident_memory_bytes+%3E+4GiB%2C+process_resident_memory_anon_bytes+%3E+4GiB%29&g0.range_input=30m&g0.end_input=2024-04-19T15%3A42%3A11&g0.tab=0&g0.relative_time=last_30_minutes&g0.tenantID=0).

If you want stripping metric names from the results, then use [label_del](https://docs.victoriametrics.com/metricsql/#label_del) function for deleting metric names before joining results:

```
union(
  metric_del(cpu_count > 10, "__name__"),
  metric_del(mem_bytes > 100G, "__name__"),
)
```

Note that in this case `mem_bytes` series are returned only if they have distinct set of [labels](https://docs.victoriametrics.com/keyconcepts/#labels) comparing to `cpu_count` series, since it is impossible to distinguish two [time series](https://docs.victoriametrics.com/keyconcepts/#time-series) with the same set of labels.

If you want filling gaps at `cpu_count > 10` results with the `mem_bytes > 100G` results, then use [`or` operator](https://prometheus.io/docs/prometheus/latest/querying/operators/#logical-set-binary-operators) instead of `union()` function:

```metricsql
(cpu_count > 10) or (mem_bytes > 100G)
```

Closing the question as answered.
